### ISSUE TITLE
NaNs during training with `tf.contrib.rnn.LayerNormBasicLSTMCell`

### ISSUE THREAD

**Issue by alanhdu** [tensorflow/tensorflow#12512]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: 1.3
- **Python version**:  3.5
- **CUDA/cuDNN version**: CUDA 8.0 and CuDNN 6 (although I can replicate without a GPU)
- **GPU model and memory**: Nvidia K80 (from Amazon P2 instance)
- **Exact command to reproduce**:

### Describe the problem

When training a model using `tf.contrib.rnn.LayerNormBasicLSTMCell`, sometimes my weights go to `nan`, even though the training data looks perfectly innocent.

~~I have **not** seen this with Tensorflow 1.2.1, which leads me to suspect that there's been a regression somewhere, but I could've just been luckier (🍀) with TF 1.2~~ (nevermind -- I have reproduced this with TF 1.2.1)

### Source code / logs

I've created two examples of this in https://github.com/alanhdu/tensorflow-12512 (clone the repo, enter a folder, and run `test.py` or build and run the `Dockerfile`. The key line(s) there are:

```python
print([np.isfinite(v).all() for v in sess.run(tf.trainable_variables())])
sess.run(train_step, feed_dict)
print([np.isfinite(v).all() for v in sess.run(tf.trainable_variables())])
```

The first print statement prints all `True`s, which is good -- but after the training step, suddenly some of the weights have `nan`s in them (and hence there's one `False` in the second print statement).

**Comment by alanhdu** [comment]
Hm... after slogging through `tfdbg`, it looks like this particular example might just be an overflow thing (although the gradient seems weirdly high and I'm not sure why this didn't happen with TF 1.2.1)

```python
import numpy as np
import tensorflow as tf
from tensorflow.python.training import training_ops

grad = np.array(
    [
        -7.22536593e+16, -2.11869319e+16, 1.31751609e+19, 6.53208027e+19,
        8.03179962e+19, 2.80957841e+16, 1.08204829e+16, -5.71410878e+15,
        -1.96436248e+15, 2.08896901e+16, -9.15358559e+19, 1.59031152e+19,
        -1.18680801e+19, -8.42016120e+19, 4.48369031e+16, 2.01818666e+16,
        3.37646465e+15, 1.25098911e+16, 2.19448051e+19, 1.66769521e+16,
        3.23241365e+16, 4.21188462e+18, -4.90144856e+18, 2.47123573e+19,
        2.10178466e+15, 1.37156984e+19, -1.02427969e+16, -6.87006047e+19,
        -8.75682902e+19, 3.87922260e+19, 7.33888846e+15, -4.27278298e+15,
        -6.55280261e+15, -1.24261527e+19, -2.42708770e+14, 1.20307027e+16,
        -7.98193545e+19, 5.26116710e+19, 1.41174265e+16, 1.92888710e+19,
        -3.83293183e+14, 1.04124438e+16, -7.68158516e+18, 3.37947113e+15,
        1.91912045e+16, -6.33320791e+15, -4.76087407e+15, 6.54729217e+15,
        2.32993309e+19, -4.43404496e+19, 3.57337199e+16, 4.92618898e+15,
        2.30337893e+16, -6.57711578e+16, -6.21250215e+15, 3.39375055e+15,
        4.12803196e+19, 4.95062884e+16, 4.91424489e+16, -1.25883009e+19,
        6.75019995e+16, -1.07368472e+20, -1.72674288e+19, -4.15710970e+18,
        5.90043635e+19, 2.97534074e+15, 9.87953683e+19, -1.51910824e+16,
        9.75750255e+16, 2.14883079e+15, 9.92727718e+18, -5.58309663e+19,
        6.07305369e+16, 4.11130619e+19, -2.76947914e+19, 1.46329901e+16,
        9.57690131e+18, 3.04936960e+15, 3.83920426e+16, 7.89846668e+19,
        -2.52319529e+18, 4.89968624e+16, 2.77158535e+15, -9.70234459e+18,
        7.63919392e+19, -7.99243622e+19, 7.39167363e+19, 1.37408169e+16,
        -5.07268442e+19, 1.93819755e+19, -1.04015647e+16, -5.94786560e+15,
        7.09745708e+15, -5.47805236e+19, -2.97251429e+19, 1.26310253e+16
    ],
    dtype=np.float32)

var = tf.get_variable("var", [96], initializer=tf.zeros_initializer)
m = tf.get_variable("m", [96], initializer=tf.zeros_initializer)
v = tf.get_variable("v", [96], initializer=tf.zeros_initializer)

beta1_power = tf.constant(0.8999999761581421)
beta2_power = tf.constant(0.9990000128746033)
lr = tf.constant(0.0010000000474974513)
beta1 = tf.constant(0.8999999761581421)
beta2 = tf.constant(0.9990000128746033)
epsilon = tf.constant(9.99999993922529e-09)

op = training_ops.apply_adam(var, m, v, beta1_power, beta2_power, lr, beta1,  beta2, epsilon, grad)
sess = tf.Session()
sess.run(tf.global_variables_initializer())
sess.run(op)
print(sess.run(var))   # see the NaNs
```

There's definitely another problem lurking here though, because I've also seen `NaN`s with gradient clipping (which should prevent the overflow) -- I'll report back with more details.

**Comment by alanhdu** [comment]
I've updated https://github.com/alanhdu/tensorflow-12512  to include two test cases (in two different folders): the `overflow` one corresponds to what I have above. The `clipped ` test case gives me NaNs even with gradient clipping, so it shouldn't be the same overflow problem.

**Comment by ali01** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by alanhdu** [comment]
@ali01 I'm happy to create a question on StackOverflow too, but:

I'm using built-in layers from Tensorflow `tf.layers.dense` and `tf.contrib.rnn.LayerNormBasicLSTMCell` on perfectly normal data and am getting garbage output (all `NaN`s). How is that not a bug? If there's a numerical stability problem here, it almost certainly exists inside the Tensorflow codebase.

**Comment by jintaojintao** [comment]
did you solve this problem now?

**Comment by AzizCode92** [comment]
I have the same issue with tf.contrib.rnn.LayerNormBasicLSTMCell and the loss is NAN and the model's performance is worse.
